arrows/qt/image_container.cxx

```diff
@@ -0,0 +1,388 @@
+size_t
+depth( QImage::Format format )
+{
+  switch ( format )
+  {
+    case QImage::Format_Mono:
+    case QImage::Format_MonoLSB:
+    case QImage::Format_Grayscale8:
+    case QImage::Format_Alpha8:
+      return 1;
+    case QImage::Format_Indexed8:
+    case QImage::Format_RGB32:
+    case QImage::Format_RGB16:
+    case QImage::Format_RGB666:
+    case QImage::Format_RGB555:
+    case QImage::Format_RGB888:
+    case QImage::Format_RGB444:
+    case QImage::Format_RGBX8888:
+    case QImage::Format_BGR30:
+    case QImage::Format_RGB30:
+      return 3;
+    case QImage::Format_ARGB32:
+    case QImage::Format_ARGB32_Premultiplied:
+    case QImage::Format_ARGB8565_Premultiplied:
+    case QImage::Format_ARGB6666_Premultiplied:
+    case QImage::Format_ARGB8555_Premultiplied:
+    case QImage::Format_ARGB4444_Premultiplied:
+    case QImage::Format_RGBA8888:
+    case QImage::Format_RGBA8888_Premultiplied:
+    case QImage::Format_A2BGR30_Premultiplied:
+    case QImage::Format_A2RGB30_Premultiplied:
+      return 4;
+    default:
+      throw kwiver::vital::image_type_mismatch_exception{
+              MODULE "::depth: unsupported image format " +
+              std::to_string( format ) };
+  }
+}
+
+// ----------------------------------------------------------------------------
+uint8_t
+get_pixel_gray( uint8_t const* in, ptrdiff_t pstep )
+{
+  Q_UNUSED( pstep );
+  return *in;
+}
+
+// ----------------------------------------------------------------------------
+uint32_t
+get_pixel_rgb( uint8_t const* in, ptrdiff_t pstep )
+{
+  return qRgb( *( in + ( 0 * pstep ) ),
+               *( in + ( 1 * pstep ) ),
+               *( in + ( 2 * pstep ) ) );
+}
+
+// ----------------------------------------------------------------------------
+uint32_t
+get_pixel_rgba( uint8_t const* in, ptrdiff_t pstep )
+{
+  return qRgba( *( in + ( 0 * pstep ) ),
+                *( in + ( 1 * pstep ) ),
+                *( in + ( 2 * pstep ) ),
+                *( in + ( 3 * pstep ) ) );
+}
+
+// ----------------------------------------------------------------------------
+template < typename T >
+QImage
+vital_to_qt( kwiver::vital::image const& img, QImage::Format format,
+             T ( * get_pixel )( uint8_t const*, ptrdiff_t ) )
+{
+  auto const w = static_cast< int >( img.width() );
+  auto const h = static_cast< int >( img.height() );
+  auto const pstep = img.d_step();
+
+  auto* const in = reinterpret_cast< uint8_t const* >( img.first_pixel() );
+  auto out = QImage{ w, h, format };
+
+  // Iterate over scanlines
+  for ( auto const j : iota( h ) )
+  {
+    auto* const inl = in + ( j * img.h_step() );
+    auto* const outl = reinterpret_cast< T* >( out.scanLine( j ) );
+
+    // Iterate over pixels in scanline
+    for ( auto const i : iota( w ) )
+    {
+      auto* const inp = inl + ( i * img.w_step() );
+      *( outl + i ) = get_pixel( inp, pstep );
+    }
+  }
+
+  return out;
+}
```

Comment:
We should consider adding the option for a shallow copy along with a deep copy. Maybe `vital_to_qt_{shallow,deep}()`.

Reply:
I specifically decided *not* to do that, both for simplicity (unlike `vil_image_view`, `QImage` cannot support the full set of formats and layouts supported by `vital::image`), and because the main use case for doing so is to be able to swap back and forth between `QImage` and `vital::image` while retaining the same underlying memory, which is not possible with `QImage` as there is no way to recover the original memory.

With `vil_image_view`, it is possible to convert to or from `vital::image` and recover the original object after a round trip. With `QImage`, each round trip would create an additional "cleanup wrapper"; you could wind up with a `QImage` wrapping a `vital::image` wrapping a `QImage` wrapping a `vital::image` wrapping a ... ad infinitum. Always making a copy prevents such a chain from forming.